frame/pallet-account/src/pallet/mod.rs

```diff
@@ -0,0 +1,228 @@
+        type CustomOriginKind: Parameter + OriginDeriving<Self::AccountId, Self::CustomOrigin>;
+        /// The runtime origin type.
+        type RuntimeOrigin: From<Self::CustomOrigin>
+            + From<frame_system::RawOrigin<Self::AccountId>>;
+        /// The overarching call type.
+        type RuntimeCall: Parameter
+            + Dispatchable<RuntimeOrigin = <Self as Config>::RuntimeOrigin>
+            + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::RuntimeCall>;
+        /// Meta transaction chain magic prefix. Required to prevent tx replay attack.
+        type ChainMagic: Get<u16>;
+        /// Meta transaction signature type.
+        type Signature: Parameter + Verify<Signer = Self::Signer>;
+        /// Meta transaction signer type.
+        type Signer: IdentifyAccount<AccountId = Self::AccountId>;
+        /// General event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Origin with given index not registered.
+        UnregisteredOrigin,
+        /// Signer not match signature, check nonce, magic and try again.
+        BadSignature,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        NewOrigin {
+            account: T::AccountId,
+            origin: T::CustomOrigin,
+        },
+        ProxyCall {
+            payer: T::AccountId,
+            origin: T::CustomOrigin,
+            result: DispatchResult,
+        },
+        MetaCall {
+            payer: T::AccountId,
+            origin: T::AccountId,
+            result: DispatchResult,
+        },
+    }
+
+    #[pallet::origin]
+    pub type Origin<T> = <T as Config>::CustomOrigin;
+
+    /// Account origins
+    #[pallet::storage]
+    pub type AccountOrigin<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, Vec<T::CustomOrigin>, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Derive new origin for account.
+        ///
+        /// The dispatch origin for this call must be _Signed_.
+        #[pallet::weight(T::WeightInfo::new_origin())]
+        #[pallet::call_index(0)]
+        pub fn new_origin(
+            origin: OriginFor<T>,
+            origin_kind: T::CustomOriginKind,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            let mut origins = AccountOrigin::<T>::get(who.clone());
+            let next_index = origins.len();
+            let new_origin = origin_kind.derive(&who, next_index as u32);
+
+            origins.push(new_origin.clone());
+            AccountOrigin::<T>::insert(who.clone(), origins);
```

Comment:
There should be a price for this - I'd say storage rent but given it's not a _rent_ since derived accounts cannot be deleted, it's more appropriate to just charge users some amount (and deposit it into treasury or burn it).

Reply:
@akru This comment is still relevant.